### ISSUE TITLE
[clang-tidy]: `bugprone-unchecked-optional-access` false-positive when modifying optional field via setter method (by reference)

### ISSUE THREAD

**Issue by BaLiKfromUA** [llvm/llvm-project#161005]
Hi folks!

Recently I found interesting bug related to accessing optional field via getter.

Initially I thought that it might be in the same nature as https://github.com/llvm/llvm-project/pull/128437 but now I think it's much more complicated.

Let's assume we have next code:
```cpp
class A {
private:
  std::optional<int> d_field;

public:
  const std::optional<int>& getField() const {
        return d_field;
  } 

  std::optional<int>& setField() {
        return d_field;
  } 

private:
  // no violations
  void do_something_with_optional() {
    if (!d_field.has_value()) {
        d_field = 42;
    }

    std::cout << d_field.value();
  }
};
```

Code below triggers warnings from check even though it's technically similar to `A::do_something_with_optional()`:
```cpp
// false positive
void foo1(A data) {
    if (!data.getField().has_value()) {
        data.setField() = 42;
    }

    std::cout << data.getField().value();
}
``` 

It also can be reproduced even by code like this:
```cpp
// false positive
void foo2(A data) {
    data.setField() = 42;
    std::cout << data.getField().value();
}
```

## My understanding why it doesn't work

IIRC the implementation of a given check, we don't analyze the access to non-const methods, even if they return optional field by reference. Plus, any access to them invalidate the stored state of related optional field that we have seen before.

Probably the fix is not simple since we cannot just rely on the method name (as I did in my previous fix), we need to understand that it returns exactly the same optional field as was modified before.

I believe the proper fix also might address the issue like https://github.com/llvm/llvm-project/issues/151287 when we call non-const method which doesn't modify optional data, but maybe I am wrong :)

I am happy to explore the solution but need an initial guidance.

Thanks!

[Compiler explorer link with my experiments](https://godbolt.org/z/Kqoj73v1d)

**Comment by BaLiKfromUA** [comment]
cc @jvoung because they helped me with previous "getter methods"-related problem :)

**Comment by BaLiKfromUA** [comment]
Thank you for the response!

I will try to make some experiments to get more data insights about this issue.

I agree about 
> I don't think this would be an easy problem to solve well. The dataflow framework does have some "interpreter" to look into e.g., 1 level of call, but I think it was too expensive to do that always.

But maybe we can start supporting some basic patterns as an opt-in functionality with performance impact.

Also will investigate if it's possible to automate proposed workaround from https://clang.llvm.org/extra/clang-tidy/checks/bugprone/unchecked-optional-access.html#stabilize-function-results

Let's keep this issue opened for now in case of any updates from my side.
